ARROW-12921: [C++][Dataset] Add RadosParquetFileFormat to Dataset API

JayjeetAtGithub opened the pull request:
The implementation includes a new `RadosParquetFileFormat` class that inherits from the `ParquetFileFormat` class to defer the evaluation of scan operations on a Parquet dataset to a RADOS storage backend. This new file format plugs into the `FileSystemDataset` API, converts filenames to object IDs using FS metadata and uses the [librados](https://docs.ceph.com/en/latest/rados/api/librados-intro/) C++ library to execute storage side functions that scan the files on the [Ceph](https://ceph.io) storage nodes (OSDs) using Arrow libraries. We ship unit and integration tests with our implementation where the tests are run against a single-node Ceph cluster.

The storage-side code is implemented as a RADOS CLS (object storage class) using [Ceph's Object Class SDK](https://docs.ceph.com/en/latest/rados/api/objclass-sdk/#:~:text=Ceph%20can%20be%20extended%20by,object%20classes%20within%20the%20tree.). The code lives in `cpp/src/arrow/adapters/arrow-rados-cls`, and is expected to be deployed on the storage nodes (Ceph's OSDs) prior to operating on tables via the `RadosParquetFileFormat` implementation. This PR includes a CMake configuration for building this library if desired (`ARROW_CLS` CMake option). We have also added Python bindings for our C++ implementations and added integration tests for them.

westonpace commented on cpp/src/arrow/dataset/file_rados_parquet.h:
```diff
@@ -0,0 +1,182 @@
+
+  Status Connect() {
+    if (rados->init2(ctx.user_name.c_str(), ctx.cluster_name.c_str(), 0))
+      return Status::Invalid("librados::init2 returned non-zero exit code.");
```
It might be simpler if `RadosInterface` returned `Status` instead of `int`.  For example, the compression_xyz.cc files convert from libxyz to Status.  You could also follow the pattern used in hdfs where `hdfs_internal.h` returns int error codes but then it is hidden by `hdfs.h` which exposes everything with `Status`/`Result`.

westonpace commented on cpp/src/arrow/dataset/rados.cc:
```diff
@@ -0,0 +1,75 @@
+// Licensed to the Apache Software Foundation (ASF) under one
```
This file (and its header) should maybe be in `src/arrow/util` or `src/arrow/filesystem`.

JayjeetAtGithub replied:
Maybe we can put this in an internal file like `file_skyhook_internal.h` which would contain this along with the RadosConnect interface ?

westonpace replied:
That could work although it wouldn't be my first conclusion.  Elsewhere the file_xyz files are about bridging the gap between a specific format (which has readers and writers in an `arrow/parquet` or `arrow/csv` or `arrow/ipc` directory and these formats know nothing of datasets) and the generic datasets API (e.g. file_base, dataset.h, etc.  These common files know nothing about any specific format).

So my thinking was more that you might not have enough in just one file to create an `arrow/rados` directory but the format-specific files should go somewhere other than `arrow/dataset`.  At a quick inventory we have:

Types that help when interacting with rados as a client (in rados.cpp/rados.h, these are analogous to our hdfs or s3 client libraries which live in `src/arrow/filesystem`.  I think `RadosConnection` and `SkyhookDirectObjectAccess` belong here as well (this might have been your point).  These types can stand on their own as Ceph integration types even if skyhook doesn't exist.

Types that define the "skyhook protocol" (the flatbuffer files, the global/static methods `SerializeScanRequest`, `DeserializeScanRequest`, `SerializeTable`, and `DeserializeTable`).  Personally, I think these might be best to live in the same directory as the cls.

Types that bridge the dataset API and the skyhook API (this would be `SkyhookFileFormat` and it uses types from both of the above categories).

westonpace commented on cpp/src/arrow/dataset/rados.h:
```diff
@@ -0,0 +1,132 @@
+
+#pragma once
+
+#include <rados/librados.hpp>
```
Ideally this would be included from the `.cc` file any any instance to `ceph::bufferlist` would be replaced by Arrow buffers or arrays.  This helps keep rados out of the Arrow public API (even if the Arrow dist was built with RADOS support).  For example, this is done with our compression utilities and filesystem utilities.  Internal headers can be used if you need templates (though I don't see any here).

westonpace commented on format/ScanRequest.fbs:
```diff
@@ -0,0 +1,33 @@
+// Licensed to the Apache Software Foundation (ASF) under one
```
This directory includes files that are meant to reflect the column format (e.g. types that all implementations are standardized on).  I'm not sure if your intention is to make the scan request a part of the specification or not.

However, I think this was probably placed here so it would get picked up by thrift.  I'm not sure what a good approach to take would be and I'll defer to future reviewers here.

JayjeetAtGithub replied:
Okay, let's wait. In general, the intention is to add a generic flatbuffer schema that can serialize some scan request parameters to be sent through the network.

westonpace replied:
So here is my current understanding.  Let me know if this seems off.  There are two pieces to this.

There is a ceph object class (called Skyhook?) which processes scan tasks and lives in a "contrib" directory.

There is a fragment / file format for Arrow that understands how to send scan requests to a ceph storage server in the skyhook format.

These two components aren't tightly coupled.  The only source of agreement is the Arrow columnar format and this flatbuffers file.  So for example (these are thought exercises, not things that will necessarily ever happen):

* Ceph could be running an older version of Skyhook built with Arrow version X and the dataset client could be running a newer version of Arrow version X+N.
* Skyhook could switch to some other library entirely in the future and as long as it continued to respect the flatbuffers format it would continue to work.
* A different non-arrow library (or an Arrow implementation in a different language) could decide to start sending requests to Skyhook and as long as they agreed upon the flatbuffers and arrow columnar format everything would continue to work.

Given the above I think the proper place for this flatbuffers file to live is in the same directory as the ceph object class.  This flatbuffers file is the API for skyhook.

Then, for building everything, the make files for that directory could produce two artifacts: A ceph object class and a small C++ "client library" which is just the output of the flatbuffers compiler.

Or you could skip the "client library" step and add an extra build step for the datasets module which runs the flatbuffers compiler.

westonpace commented on python/pyarrow/_rados.pxd:
```diff
@@ -0,0 +1,38 @@
+# Licensed to the Apache Software Foundation (ASF) under one
```
I wonder if we need a brand new module here or not.  For example, it could be in the dataset module and then it could be an error at use time if someone tries to use it.  Currently the parquet file format and csv file format live in the dataset module (even though there is already a parquet module and csv module).  However, those two are generally guaranteed to be installed.  I'm not sure which pattern to follow for rados.

westonpace commented on cpp/src/arrow/adapters/arrow-rados-cls/README.md:
```diff
@@ -0,0 +1,78 @@
+-->
+# <img src="https://iris-hep.org/assets/logos/skyhookdmLogoJeff.png" width="64" valign="middle" alt="Skyhook"/> SkyhookDM-Arrow
+
+Apache Arrow provides a [`Dataset`](https://arrow.apache.org/docs/cpp/api/dataset.html) API, which acts as an abstraction over a collection of files in different storage backends like S3 and HDFS. It supports different file formats like CSV and Parquet through the [`FileFormat`](https://arrow.apache.org/docs/cpp/api/dataset.html#_CPPv4N5arrow7dataset10FileFormatE) API. In SkyhookDM, we create a n ...
```
I think we should probably avoid references to Skyhook from within Arrow.  I think that will raise confusion.  Could we describe the architecture with something more like "Arrow Ceph object class" terminology?

westonpace commented on cpp/src/arrow/adapters/arrow-rados-cls/scripts/splitter.py:
```diff
@@ -0,0 +1,32 @@
+# Licensed to the Apache Software Foundation (ASF) under one
```
I think this is more of a "skyhook" utility and less of a "Arrow utility".  It maybe shouldn't be included in this PR.

JayjeetAtGithub replied:
Fixed

westonpace commented on cpp/src/arrow/adapters/arrow-rados-cls/docs/architecture.md:
```diff
@@ -0,0 +1,41 @@
+<!---
```
This document is very good for helping to understand the intent of the PR but I'm not sure if we want this, as written, to be merged in.  I don't think your goal is to embed the Skyhook project inside of Arrow correct?  I think the goal is only to include the Arrow-Rados integration (and the Arrow Ceph object class)?  I may be misunderstanding things too.

carlosmalt replied:
@westonpace Thank you very much for your review. Since the beginning of this year the Skyhook project _is_ the Arrow-RADOS integration. They have become the same because we discovered that Arrow had already all the parts we needed.

carlosmalt replied:
@westonpace After discussing this with @JayjeetAtGithub, I confirm that the goal is to only include the Arrow-RADOS integration (including the Arrow Ceph object class). Jayjeet will update [the architecture document](https://github.com/uccross/skyhookdm-arrow/blob/rados-parquet-pr/cpp/src/arrow/adapters/arrow-rados-cls/docs/architecture.md) to highlight the parts that will be part of this pull request. The current plan is to put all other parts into a separate repository that will have Arrow as a submodule.

westonpace reviewed (changes requested):
This is very cool and it looks like a lot of great work.  Thank you for putting this together.  It adds some good capabilities and examples of distributing work with Arrow.  I'm still looking into this and will follow up with more detailed comments but I've got some coarse initial thoughts to start the discussion:

* The `cpp/src/arrow/adapters` folder should probably be `cpp/src/adapters/ceph` (or possibly even `adapters/ceph/cpp/src` since it is a standalone module (that depends on Arrow) and not an extension/plugin into Arrow itself.
* The approach of extending ParquetFileFormat seems like it might lead to duplicate code since you'd need a RadosXyzFileFormat for every fragment.  Instead would it be possible to have RadosFragment which takes in (via the constructor) a shared_ptr to a FileFormat?  The FileFormat could then be serialized (which format to use and read options) and sent as part of  the rados scan request.

JayjeetAtGithub commented:
@westonpace Thanks a lot for taking the time to review the PR. I wanted to let you know that I have just updated the Pull request branch with the latest changes we had in our development branch (in our fork). The main change that you would see now as compared to the last time you looked at the PR, is that we changed the `RadosParquetFileFormat` name to `SkyhookFileFormat`, which would act as an abstraction to offload into Ceph any real file format supported by Arrow. Currently, `SkyhookFileFormat` supports LZ4 compressed IPC (feather) and Parquet.

JayjeetAtGithub commented on ci/scripts/cpp_build.sh:
```diff
@@ -85,6 +85,8 @@ cmake -G "${CMAKE_GENERATOR:-Ninja}" \
       -DARROW_RUNTIME_SIMD_LEVEL=${ARROW_RUNTIME_SIMD_LEVEL:-MAX} \
       -DARROW_S3=${ARROW_S3:-OFF} \
+      -DARROW_RADOS=${ARROW_RADOS:-OFF} \
+      -DARROW_CLS=${ARROW_CLS:-OFF} \
```
@westonpace Since the API's we added are named as Skyhook**, can we name the flag as `ARROW_SKYHOOK` ?

westonpace replied:
I'm pretty sure I understand where you are coming from.  Thinking on this more I think it boils down to the idea of hosting a separate tool (Skyhook) in Arrow.  I'm going to raise this on the mailing list for clarification as I'm not sure we have a clear precedent.

westonpace commented on cpp/src/arrow/adapters/arrow-rados-cls/cls_arrow.cc:
```diff
@@ -0,0 +1,292 @@
+
+  // Scan the object
+  std::shared_ptr<arrow::Table> table;
+  if (file_format == 0) {
```
You might want to consider an enum here instead of magic numbers.

westonpace commented on cpp/src/arrow/adapters/arrow-rados-cls/cls_arrow.cc:
```diff
@@ -0,0 +1,292 @@
+                                                   &projection_schema, &dataset_schema,
+                                                   file_size, file_format, *in))
+           .ok()) {
+    CLS_LOG(0, "error: %s", s.message().c_str());
```
Maybe extract this into a helper method `LogArrowError` or something?

westonpace reviewed (commented):
I hope to finish this first pass on Monday but here is a new batch of comments in the meantime.

JayjeetAtGithub commented on .github/workflows/docs.yml:
```diff
@@ -0,0 +1,57 @@
+# # Licensed to the Apache Software Foundation (ASF) under one
```
Done

JayjeetAtGithub commented:
> I hope to finish this first pass on Monday but here is a new batch of comments in the meantime.

Thanks a lot

JayjeetAtGithub commented on .travis.yml:
```diff
@@ -79,6 +79,44 @@ jobs:
         ARCH: arm64v8
         ARROW_CI_MODULES: "GO"
         DOCKER_IMAGE_ID: debian-go
+    
```
Hi @westonpace, The issues (S3 and no-cache) were a problem in an earlier version of Arrow and the comments are also from an old version of arrow. I believe the changes somehow got carried forward in my PR. I believe they have been already addressed.

JayjeetAtGithub commented on cpp/src/arrow/adapters/arrow-rados-cls/cls_arrow.cc:
```diff
@@ -0,0 +1,292 @@
+
+  ARROW_RETURN_NOT_OK(builder->Filter(filter));
+  ARROW_RETURN_NOT_OK(builder->Project(projection_schema->field_names()));
+  ARROW_RETURN_NOT_OK(builder->UseThreads(false));
```
We can use threads. That boosts up performance using by more CPU.

westonpace commented on cpp/src/arrow/dataset/file_skyhook.cc:
```diff
@@ -0,0 +1,280 @@
+
+SkyhookFileFormat::SkyhookFileFormat(
+    const std::shared_ptr<connection::RadosConnection>& connection) {
+  connection->Connect();
```
I'm not sure what resources are consumed when calling `Connect` but generally I wouldn't expect any connections to be opened just be creating a `SkyhookFileFormat` instance.  It might make more sense to establish the connection on `Inspect` or `ScanFile`.

westonpace commented on cpp/src/arrow/dataset/file_skyhook.cc:
```diff
@@ -0,0 +1,280 @@
+    ARROW_RETURN_NOT_OK(doa_->Exec(st.st_ino, "scan_op", request, result));
+
+    RecordBatchVector batches;
+    ARROW_RETURN_NOT_OK(DeserializeTable(batches, result, !options_->use_threads));
```
Why `!options_->use_threads`?

JayjeetAtGithub replied:
I did this, because earlier arrow didn't support nested threading very well, and what used to happen was threaded decompression caused the execution to enter an deadlocked state.

JayjeetAtGithub replied:
Looks like the decompression API also needs to be futurized for the async scanner to work. Since, it is not yet futurized,  looks like we need to do this check to avoid nested threading.

westonpace commented on cpp/src/arrow/dataset/file_skyhook.h:
```diff
@@ -0,0 +1,275 @@
+  std::string ConvertFileInodeToObjectID(uint64_t inode) {
+    std::stringstream ss;
+    ss << std::hex << inode;
+    std::string oid(ss.str() + ".00000000");
```
Maybe add a comment or some explanation behind this magic string?

westonpace commented on cpp/src/arrow/dataset/file_skyhook_test.cc:
```diff
@@ -0,0 +1,95 @@
+#include "arrow/dataset/scanner.h"
+#include "arrow/dataset/test_util.h"
+
+#define ABORT_ON_FAILURE(expr)                     \
```
I think we have `ASSERT_OK` (for tests),  `ABORT_NOT_OK` (for scripts and places where error checking isn't allowed) and `DCHECK` (for debug assertions).  I'm not sure we need a new macro.

westonpace commented on cpp/src/arrow/dataset/scanner_internal.h:
```diff
@@ -185,6 +185,12 @@ inline Result<ScanTaskIterator> GetScanTaskIterator(
   auto fn = [options](std::shared_ptr<Fragment> fragment) -> Result<ScanTaskIterator> {
     ARROW_ASSIGN_OR_RAISE(auto scan_task_it, fragment->Scan(options));
 
+    if (fragment->type_name() == "skyhook") {
```
Is Skyhook actually doing projection?  I think filtering is fine but if projection is done early at the format level then I believe that may cause issues when executing datasets that come from a variety of sources (e.g. unions and joins).

lidavidm replied:
Or rather, we shouldn't skip filter and projection here right? Even if it's pushed down (as can be done already in Parquet) we should keep the rest of the pipeline uniform. (At least, projection should be cheap - it may be understandable to skip filtering entirely if the format guarantees that filtering is completely pushed down.)

JayjeetAtGithub replied:
Yes, filtering (although is quite fast) incurs some overhead. Maybe we can have a flag in `ScanOptions` to switch on/off client-side filtering ?

westonpace reviewed (commented):
A few more high level comments.  Just to reiterate, I like this high level approach and think this is a valuable addition.  Perhaps the easiest way to get this to merge would be to break it up into smaller commits.  Off the top of my head I think it could break down into three independent chunks of work.

1. Add the Ceph class and the flatbuffer files defining the skyhook protocol.  Tests could verify the Ceph class builds and installs onto a Ceph server correctly.  Testing of the skyhook protocol itself may or may not be possible at this point.
2. Add Ceph "client" types like `SkyhookDirectObjectAccess` and the types in `rados.h`.  Make sure you can stat and list files and you can make basic use of the exec API.
3. Add the skyhook fragment and bridge the dataset API and the skyhook protocol.

You don't have to do this and there are other ways you could divide the work as well.  Please don't hesitate to let me know if I can provide any help.  Feel free to reach out over email if you'd like to have a Google Meet meeting or something to talk about any details.

When you have addressed all my feedback please go ahead and re-request a review from me.

JayjeetAtGithub commented:
Thanks @westonpace for sharing your thoughts.

> So here is my current understanding. Let me know if this seems off. There are two pieces to this.
> 
> There is a ceph object class (called Skyhook?) which processes scan tasks and lives in a "contrib" directory.
> 
> There is a fragment / file format for Arrow that understands how to send scan requests to a ceph storage server in the skyhook format.
> 
That is correct.

> These two components aren't tightly coupled. The only source of agreement is the Arrow columnar format and this flatbuffers file. So for example (these are thought exercises, not things that will necessarily ever happen):
> 
> * Ceph could be running an older version of Skyhook built with Arrow version X and the dataset client could be running a newer version of Arrow version X+N.

Yeah, this could happen. In this case, we need to ensure that the storage side understand the `ScanRequest` language in which the client sends requests and can also the serialize tables in a buffer format understandable/supported by the client.

> * Skyhook could switch to some other library entirely in the future and as long as it continued to respect the flatbuffers format it would continue to work.

Similar as above I guess.

> * A different non-arrow library (or an Arrow implementation in a different language) could decide to start sending requests to Skyhook and as long as they agreed upon the flatbuffers and arrow columnar format everything would continue to work.

Yes, as along as both the client and server agrees upon the same send and receive protocol, they should work fine.

> Given the above I think the proper place for this flatbuffers file to live is in the same directory as the ceph object class. This flatbuffers file is the API for skyhook.

I agree. This is the send API for skyhook.

> Then, for building everything, the make files for that directory could produce two artifacts: A ceph object class and a small C++ "client library" which is just the output of the flatbuffers compiler.
> Or you could skip the "client library" step and add an extra build step for the datasets module which runs the flatbuffers compiler.

Could you please explain this part a little more?

JayjeetAtGithub commented on ci/scripts/integration_ceph.sh:
```diff
@@ -0,0 +1,129 @@
+
+    # download an example dataset and copy into the mounted dir
+    rm -rf nyc*
+    wget https://raw.githubusercontent.com/JayjeetAtGithub/zips/main/nyc.zip
```
Can we use a python script to generate a random dataset of parquet files instead of this `wget` here ?

westonpace commented:
> Could you please explain this part a little more?

Sure, let me prioritize and group things a bit too.

# Move flatbuffers out of `format`

The file `ScanRequest.fbs` should come out of the `format` directory.  This file is the "Skyhook" portion of the code and do not need to be a part of Arrow.

The `format` directory represents the core Arrow protocols that need to be able to stand alone in any language or on any hardware.  Since `ScanRequest.fbs` is specific to Rados/Skyhook it can't go here.  I don't think this requirement is very flexible.  Unfortunately, it does probably mean you need to add a new flatbuffers compilation step or modify the existing one to build a new directory.

# Move `cls` files out of `cpp/src/arrow`

The `cpp/src/arrow` directory (and its subdirectories) create object files meant to be included by developers developing code based on Arrow.  `cls_arrow.cc` isn't really the same thing.  It is a "standalone artifact" that is deployed on a Ceph server.  It should go in its own directory to make it clear.  I think going one directory up to `cpp/src/ceph` or `cpp/src/skyhook` would be fine.  Although, as these tools grow we may want to create a `cpp/src/contrib` (and move the other items `cpp/src/arrow/adapters` here too).

I suppose there is some precedence with the `orc`/`tensorflow` stuff here so I could be convinced this isn't a big deal but I feel it would be cleaner to keep it separate.

# Splitting code into service/protocol/client/client-adapter/dataset-adapter

For other integrations (s3, parquet, orc, csv, hdfs) the code tends to follow a certain pattern.  If you follow this pattern it will be clearer what the purpose of each of your files is.  This means moving some of your code around (e.g. you combine client code and dataset-adapter code in the same file).  This ask is lower priority than the previous asks but it will help clean things up.

**Service**: This 3rd party code represents the service itself.  This code should stand on its own (if it uses Arrow that is an implementation detail and could change with no real consequence).  For most integrations we don't actually have this code in the Arrow code base.  For example, we don't have the code to run an S3 server.  We don't have the code to run an HDFS cluster.  For formats like CSV or parquet there is no "service".

For this PR the "service" is the Ceph class.

**Protocol/API**: This 3rd party concept defines how to interact with the service.  Again, this should stand on its own and the API should be consumable by tools other than Arrow.  It may or may not be compiled into code.  For example, with parquet we have `src/parquet/parquet.thrift`.  S3 has an HTTP API but we don't interact with it directly.

For this PR this is represented by the Skyhook flatbuffer files.

**Client**: This 3rd party code makes it easier to interact with the protocol.  Again, this stands on its own and if it is using Arrow that is an implementation detail.  For S3 we have the AWS S3 API.  For HDFS we have libhdfs.  For parquet we have `cpp/src/parquet`.

This is not as clear in the PR but I think it consists of librados, `rados.h`, `rados.cc`, and the methods `SerializeScanRequest`, `DeserializeScanRequest`, `SerializeTable`, and `DeserializeTable`.

**Client Adapter**: This Arrow code wraps a client interface and completely hides any client headers.  The input and output of the client adapter should be Arrow types (although it will be converting to client types under the hood).  For S3 we have `cpp/src/arrow/filesystem/s3fs.h`.  For hdfs we have `cpp/src/arrow/filesystem/hdfs.h`.  For Parquet we have `cpp/src/parquet/arrow`.

For this PR I think this consists of `RadosConnection` and `SkyhookDirectObjectAccess` although some improvement is needed here because these types don't fully hide `ceph::bufferlist`.

**Dataset Adapter**: This Arrow-Dataset code uses the Client Adapter to create an implementation of FileFormat or Fragment to bridge the gap between the datasets API and the client service.

For this PR this consists of `SkyhookFileFormat`.

# Splitting the PR into multiple PRs

This is completely optional but I personally think it will be faster and easier to break things into smaller pieces.

I'm flexible on most things so let me know if any of this seems like it will be a problem or is overly strict.

lidavidm commented on cpp/src/arrow/dataset/file_skyhook.h:
```diff
@@ -0,0 +1,275 @@
+
+/// \brief Serialize the result Table to a bufferlist.
+/// \param[in] table The table to serialize.
+/// \param[in] aggressive If true, use ZSTD compression instead of LZ4.
```
Is there a reason to hardcode ZSTD/LZ4 instead of accepting an Arrow codec instance or codec type?

lidavidm commented on cpp/src/arrow/dataset/scanner.h:
```diff
@@ -104,6 +104,8 @@ struct ARROW_DS_EXPORT ScanOptions {
   /// Note: The IOContext executor will be ignored if use_threads is set to false
   io::IOContext io_context;
+  // Partition expression
+  compute::Expression partition_expression = compute::literal(true);
```
Instead of stuffing it into the scan options, it might be better to thread this through from SkyhookFileFormat into SkyhookScanTask and then into the serialized scan request.

lidavidm commented on cpp/src/arrow/dataset/scanner.cc:
```diff
@@ -32,6 +32,10 @@
 #include "arrow/dataset/dataset.h"
 #include "arrow/dataset/dataset_internal.h"
 #include "arrow/dataset/scanner_internal.h"
+#include "arrow/io/memory.h"
```
nit: are these includes necessary here?

JayjeetAtGithub replied:
Scanner.cc does not need any change anymore.

lidavidm reviewed (commented):
I took only a quick high-level look but in general I agree with Weston: the code here can be reorganized, and it would be appreciated (though not required) if this PR could be broken up. Additionally there's some discussion about accepting external plugins to Datasets and accepting 'contrib' modules; it might be worth considering if this could be structured as a contrib plugin.

JayjeetAtGithub commented:
@westonpace @lidavidm Thanks for taking the time to explain in detail the proper way to contribute skyhook into arrow. I am writing down some action plan that I figured out from the discussion till now. I guess I will break down this PR into 3 parts now 

i) An initial C++ implementation of skyhook (both `SkyhookFileFormat` and the object class code) placed in `cpp/src/skyhook/client` and `cpp/src/skyhook/cls`. We can place the protocol (like the ScanRequest and the helper serializer/deserialize methods) in `cpp/src/skyhook/protocol` and use it in both the client and CLS code.  Here we will build two libraries: a `libarrow_skyhook_client.so` for the client-side file format extensions and a `libcls_skyhook.so` for the CLS side Ceph object classes. Some unit/integration tests would also be included.
 
ii) The python bindings for the `SkyhookFileFormat` extension with python unit tests.
iii) Other misc. improvements to the code change to finalize it for merging.

I would like to keep this PR open as a reference to all the valuable reviews that you have given so that I can incorporate all of them in my 1st PR with the minimal C++ implementations. Thanks.

JayjeetAtGithub commented on cpp/src/arrow/dataset/discovery.cc:
```diff
@@ -270,7 +270,14 @@ Result<std::shared_ptr<Dataset>> FileSystemDatasetFactory::Finish(FinishOptions
+    std::shared_ptr<FileFragment> fragment;
+    if (format_->type_name() == "skyhook") {
+      ARROW_ASSIGN_OR_RAISE(fragment,
+                            format_->MakeFragment({info, fs_}, partition, true, schema));
```
Removed this change. Turns out this is not required anymore with dataset schema populated automatically in ScanOptions.

bkietz commented on cpp/src/arrow/dataset/file_skyhook.h:
```diff
@@ -0,0 +1,275 @@
+  std::shared_ptr<connection::RadosConnection> connection_;
+};
+
+/// \class SkyhookFileFormat
+/// \brief A ParquetFileFormat implementation that offloads the fragment
+/// scan operations to the Ceph OSDs
+class ARROW_DS_EXPORT SkyhookFileFormat : public ParquetFileFormat {
```
I think this is not the interface you want to implement for skyhook.

A `FileFormat`'s fragments should be able to wrap files in any file system- or even a buffer in memory instead of a file. However IIUC `SkyhookFileFormat` always assumes we can do direct object access, which assumes we have a path and a file system which supports direct object access.

Instead I think this should be exposed as a `Fragment` subclass (**not** a `FileFragment`), or maybe even an implementation of `Scanner` since it overrides client-side filtering and projection.

carlosmalt replied:
Thank you, @bkietz -- I agree with you that `FileFormat` is not the ideal place. I think we proposed something similar to what you suggest in @ivotron's [pull request](https://github.com/apache/arrow/pull/8647) (see also [mailing list thread](https://lists.apache.org/thread.html/r2a5a693967213b7c6bb49015194ca16afc4d20047805d0e069c2e45c%40%3Cdev.arrow.apache.org%3E)). However he ended up closing that pull request because he realized we would be re-implementing a lot of file system functionality. I've just talked to @ivotron about this and he pointed out that  the fact that we are getting bounced between these two approaches shows that the notion of offloading computation to the storage layer is a more fundamental issue than a specialization of`Scanner` or `FileFormat`. I hope that the `SkyhookFileFormat` can be merged to provide a use case for providing a more general solution to offloading in Arrow.

westonpace replied:
> the notion of offloading computation to the storage layer is a more fundamental issue than a specialization of Scanner or FileFormat

This is likely true.  It would be good to start brainstorming what sorts of concepts we are missing.  The mailing list could be a good place for discussion & brainstorming or future PRs could be a good place for more concrete concepts/proposals.

At the moment we have (I'll add that these APIs have evolved considerably since the start of the year):

 * **FileFormat** - A file format converts a stream of bytes (regardless of source) into a stream of record batches.
 * **Scanner** - A scanner converts a dataset (list of fragments) into a stream of record batches.  The scanner also applies filtering & projection.
 *  **Fragment** - An independently scannable unit
 *  **FileFragment** - An independently scannable unit backed by a file, relies on `FileFormat`.

The current scanner is also a fragment scheduler.  Given a list of fragments it selects N fragments to start scanning and merges (and sometimes sequences) their results.

Arrow itself is also in flux with these interfaces.  The scanner is being split into a scan node which does fragment scheduling and the filtering/projection pieces are being pulled into their own nodes.

The downside of skyhook being a scanner that I can see is that skyhook doesn't do any fragment scheduling today, and it would have to reinvent that.  On the other hand, maybe skyhook knows more information that would help it do this scheduling.  For example, Arrow just sort of picks "8 fragments at a time" is probably good enough (kDefaultFragmentReadahead) which is overkill for HDD, about right for SDD and S3, and underkill for certain EC2/S3 configurations.  I have no idea what is appropriate for skyhook (I assume it would be based on the # of nodes in the ceph cluster and maybe the file distribution).  This particular fact wouldn't necessarily require an additional scanner implementation but maybe some other fact does (e.g. maybe there is a certain ordering in which fragments should be scheduled).  I can understand why one wouldn't want Skyhook to be a Scanner.

The downside of skyhook being a file format is that, if it isn't the right fit, it starts to break apart the abstraction.  For example, we have to introduce `FileFormat::dataset_schema`, there are potential Liskov violations which could mean the code is more brittle (i.e. we might start using FileFormat differently or elsewhere and the Rados implementation doesn't fit the new usage).

Perhaps Skyhook is a good fit for `Fragment`.  This would allow Skyhook to take advantage of the Scanner's scheduling but give it more flexibility.  Adding `handles_filtering` and `handles_projection` properties to a fragment seems appropriate too.  Or, Skyhook could attach a "guarantee" to the returned batch (this concept isn't in scanner yet but it is emerging in the exec plan abstractions) so that future filtering mechanisms would skip it because the guarantee shows it is already filtered.

I'm also pretty amenable to the logic of "we don't need to get it perfect now" and we can get something working and iterate on it in the future.  I appreciate the effort to break things into smaller PRs.
